src/bedrock/hash_ring.cpp

```diff
@@ -0,0 +1,212 @@
+
+  for (auto it = mts.begin(); it != mts.end(); it++) {
+    string ip = it->get_ip();
+    auto tids = responsible_local(key, DEFAULT_LOCAL_REPLICATION, local_memory_hash_ring);
+
+    for (auto iter = tids.begin(); iter != tids.end(); iter++) {
+      threads.insert(server_thread_t(ip, *iter));
+    }
+  }
+
+  return threads;
+}
+
+void issue_replication_factor_request(
+    const string& respond_address,
+    const string& key,
+    global_hash_t& global_memory_hash_ring,
+    local_hash_t& local_memory_hash_ring,
+    SocketCache& pushers,
+
+    unsigned& seed) {
+  string key_rep = string(METADATA_IDENTIFIER) + "_" + key + "_replication";
+  auto threads = get_responsible_threads_metadata(key_rep, global_memory_hash_ring, local_memory_hash_ring);
+
+  if (threads.size() == 0) {
+    cerr << "error!\n";
+  }
+
+  string target_address = next(begin(threads), rand_r(&seed) % threads.size())->get_request_pulling_connect_addr();
+
+  communication::Request req;
+  req.set_type("GET");
+  req.set_respond_address(respond_address);
+
+  prepare_get_tuple(req, string(METADATA_IDENTIFIER) + "_" + key + "_replication");
+  push_request(req, pushers[target_address]);
+}
+
+// get all threads responsible for a key from the "node_type" tier
+// metadata flag = 0 means the key is a metadata. Otherwise, it is a regular data
+unordered_set<server_thread_t, thread_hash> get_responsible_threads(
+    string respond_address,
+    string key,
+    bool metadata,
+    unordered_map<unsigned, global_hash_t>& global_hash_ring_map,
+    unordered_map<unsigned, local_hash_t>& local_hash_ring_map,
+    unordered_map<string, key_info>& placement,
+    SocketCache& pushers,
+    vector<unsigned>& tier_ids,
+    bool& succeed,
+    unsigned& seed) {
+
+  if (metadata) {
+    succeed = true;
+    return get_responsible_threads_metadata(key, global_hash_ring_map[1], local_hash_ring_map[1]);
+  } else {
+    unordered_set<server_thread_t, thread_hash> result;
+
+    if (placement.find(key) == placement.end()) {
+      issue_replication_factor_request(respond_address, key, global_hash_ring_map[1], local_hash_ring_map[1], pushers, seed);
+      succeed = false;
+    } else {
+      for (auto id_iter = tier_ids.begin(); id_iter != tier_ids.end(); id_iter++) {
+        unsigned tier_id = *id_iter;
+        auto mts = responsible_global(key, placement[key].global_replication_map_[tier_id], global_hash_ring_map[tier_id]);
+
+        for (auto it = mts.begin(); it != mts.end(); it++) {
+          string ip = it->get_ip();
+          auto tids = responsible_local(key, placement[key].local_replication_map_[tier_id], local_hash_ring_map[tier_id]);
+
+          for (auto iter = tids.begin(); iter != tids.end(); iter++) {
+            result.insert(server_thread_t(ip, *iter));
+          }
+        }
+      }
+
+      succeed = true;
+    }
+
+    return result;
+  }
+}
+
+
+// query the routing for a key and return all address
+vector<string> get_address_from_routing(
```

Comment:
What's the reasoning behind putting this function in hash_ring.cpp instead of common.cpp?

Reply:
Seemed like it wasn't very "common", but pretty hash_ring-ish. More specifically it needs the definition of user threads and I tried to pull the thread stuff out of common.cpp.